### ISSUE TITLE
False positive syntax error in range pattern

### ISSUE THREAD

**Issue by lf-** [rust-lang/rust-analyzer#9779]
This program, which rustc compiles fine, generates a syntax error according to rust-analyzer:

```rust
fn main() {
    let _ = match 2u32 {
        0 => 1,
        1.. => 5,
        //  ^^^ Syntax Error: expected pattern
        //  [rust-analyzer: syntax-error]
    };
}
```

<details><summary>`rust-analyzer parse` output</summary>
<pre><code>
dev/rust-analyzer - [master●] » cargo r --release -p rust-analyzer -- parse < oops.rs
    Finished release [optimized] target(s) in 0.06s
     Running `target/release/rust-analyzer parse`
SOURCE_FILE@0..80
  FN@0..79
    FN_KW@0..2 "fn"
    WHITESPACE@2..3 " "
    NAME@3..7
      IDENT@3..7 "main"
    PARAM_LIST@7..9
      L_PAREN@7..8 "("
      R_PAREN@8..9 ")"
    WHITESPACE@9..10 " "
    BLOCK_EXPR@10..79
      L_CURLY@10..11 "{"
      WHITESPACE@11..16 "\n    "
      LET_STMT@16..77
        LET_KW@16..19 "let"
        WHITESPACE@19..20 " "
        WILDCARD_PAT@20..21
          UNDERSCORE@20..21 "_"
        WHITESPACE@21..22 " "
        EQ@22..23 "="
        WHITESPACE@23..24 " "
        MATCH_EXPR@24..76
          MATCH_KW@24..29 "match"
          WHITESPACE@29..30 " "
          LITERAL@30..34
            INT_NUMBER@30..34 "2u32"
          WHITESPACE@34..35 " "
          MATCH_ARM_LIST@35..76
            L_CURLY@35..36 "{"
            WHITESPACE@36..45 "\n        "
            MATCH_ARM@45..52
              LITERAL_PAT@45..46
                LITERAL@45..46
                  INT_NUMBER@45..46 "0"
              WHITESPACE@46..47 " "
              FAT_ARROW@47..49 "=>"
              WHITESPACE@49..50 " "
              LITERAL@50..51
                INT_NUMBER@50..51 "1"
              COMMA@51..52 ","
            WHITESPACE@52..61 "\n        "
            MATCH_ARM@61..67
              RANGE_PAT@61..66
                LITERAL_PAT@61..62
                  LITERAL@61..62
                    INT_NUMBER@61..62 "1"
                DOT2@62..64 ".."
                WHITESPACE@64..65 " "
                ERROR@65..66
                  EQ@65..66 "="
              ERROR@66..67
                R_ANGLE@66..67 ">"
            WHITESPACE@67..68 " "
            MATCH_ARM@68..70
              LITERAL_PAT@68..69
                LITERAL@68..69
                  INT_NUMBER@68..69 "5"
              ERROR@69..70
                COMMA@69..70 ","
            WHITESPACE@70..75 "\n    "
            R_CURLY@75..76 "}"
        SEMICOLON@76..77 ";"
      WHITESPACE@77..78 "\n"
      R_CURLY@78..79 "}"
  WHITESPACE@79..80 "\n"
</code></pre>
</details>

Rust analyzer version: 8a843113

**Comment by lf-** [comment]
```
dev/rust-analyzer - [master●] » RUSTC_BOOTSTRAP=1 rustc +stable oops.rs
error[E0658]: exclusive range pattern syntax is experimental
 --> oops.rs:4:9
  |
4 |         1..u32::MAX => 5,
  |         ^^^^^^^^^^^
  |
  = note: see issue #37854 <https://github.com/rust-lang/rust/issues/37854> for more information
  = help: add `#![feature(exclusive_range_pattern)]` to the crate attributes to enable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
```

Ahh it's just been stabilized... Well that explains part of it. I will have a go at figuring out why rust analyzer doesn't like it anyway. I'm shaving yaks today.

**Comment by Zij-IT** [comment]
I believe this issue is still around. I am currently on rust-analyzer 2021-10-18 stable, and am experiencing the issue with the following snippet. I have commented where the errors arise. I would be happy to provide any additional information, just let me know what you need! 

```
match (1, 2, 3) {
    (1.., 4.., 3..) => {}, // Syntax Error: expected pattern
    (_, _, _) => {},
}

match (1, 2) {
    (1.., 4..=i32::MAX) => {}, // Syntax Error: expected pattern
    (_, _) => {},
}
```

These snippets do not error:

```
match (1, 2, 3) {
    (1..=i32::MAX, 4..=i32::MAX, 3..) => {},
    (_, _, _) => {},
}

match (1, 2) {
    (1..=i32::MAX, 4..) => {},
    (_, _) => {},
}
```

**Comment by iDawer** [comment]
> I believe this issue is still around. I am currently on rust-analyzer 2021-10-18 stable, and am experiencing the issue with the following snippet. I have commented where the errors arise. I would be happy to provide any additional information, just let me know what you need!
> 
> ```
> match (1, 2, 3) {
>     (1.., 4.., 3..) => {}, // Syntax Error: expected pattern
>     (_, _, _) => {},
> }
> 
> match (1, 2) {
>     (1.., 4..=i32::MAX) => {}, // Syntax Error: expected pattern
>     (_, _) => {},
> }
> ```



For rust-analyzer version: 88024c7ec 2022-06-03 nightly there is no error for me:

```rust
fn foo() {
    match (1, 2, 3) {
        (1.., 4.., 3..) => {},
        (1.., 4..=i32::MAX) => {},
    }
}
```

<details>
  <summary>syntax</summary>

```
SOURCE_FILE@0..106
  FN@0..106
    FN_KW@0..2 "fn"
    WHITESPACE@2..3 " "
    NAME@3..6
      IDENT@3..6 "foo"
    PARAM_LIST@6..8
      L_PAREN@6..7 "("
      R_PAREN@7..8 ")"
    WHITESPACE@8..9 " "
    BLOCK_EXPR@9..106
      STMT_LIST@9..106
        L_CURLY@9..10 "{"
        WHITESPACE@10..15 "\n    "
        MATCH_EXPR@15..104
          MATCH_KW@15..20 "match"
          WHITESPACE@20..21 " "
          TUPLE_EXPR@21..30
            L_PAREN@21..22 "("
            LITERAL@22..23
              INT_NUMBER@22..23 "1"
            COMMA@23..24 ","
            WHITESPACE@24..25 " "
            LITERAL@25..26
              INT_NUMBER@25..26 "2"
            COMMA@26..27 ","
            WHITESPACE@27..28 " "
            LITERAL@28..29
              INT_NUMBER@28..29 "3"
            R_PAREN@29..30 ")"
          WHITESPACE@30..31 " "
          MATCH_ARM_LIST@31..104
            L_CURLY@31..32 "{"
            WHITESPACE@32..41 "\n        "
            MATCH_ARM@41..63
              TUPLE_PAT@41..56
                L_PAREN@41..42 "("
                RANGE_PAT@42..45
                  LITERAL_PAT@42..43
                    LITERAL@42..43
                      INT_NUMBER@42..43 "1"
                  DOT2@43..45 ".."
                COMMA@45..46 ","
                WHITESPACE@46..47 " "
                RANGE_PAT@47..50
                  LITERAL_PAT@47..48
                    LITERAL@47..48
                      INT_NUMBER@47..48 "4"
                  DOT2@48..50 ".."
                COMMA@50..51 ","
                WHITESPACE@51..52 " "
                RANGE_PAT@52..55
                  LITERAL_PAT@52..53
                    LITERAL@52..53
                      INT_NUMBER@52..53 "3"
                  DOT2@53..55 ".."
                R_PAREN@55..56 ")"
              WHITESPACE@56..57 " "
              FAT_ARROW@57..59 "=>"
              WHITESPACE@59..60 " "
              BLOCK_EXPR@60..62
                STMT_LIST@60..62
                  L_CURLY@60..61 "{"
                  R_CURLY@61..62 "}"
              COMMA@62..63 ","
            WHITESPACE@63..72 "\n        "
            MATCH_ARM@72..98
              TUPLE_PAT@72..91
                L_PAREN@72..73 "("
                RANGE_PAT@73..76
                  LITERAL_PAT@73..74
                    LITERAL@73..74
                      INT_NUMBER@73..74 "1"
                  DOT2@74..76 ".."
                COMMA@76..77 ","
                WHITESPACE@77..78 " "
                RANGE_PAT@78..90
                  LITERAL_PAT@78..79
                    LITERAL@78..79
                      INT_NUMBER@78..79 "4"
                  DOT2EQ@79..82 "..="
                  PATH_PAT@82..90
                    PATH@82..90
                      PATH@82..85
                        PATH_SEGMENT@82..85
                          NAME_REF@82..85
                            IDENT@82..85 "i32"
                      COLON2@85..87 "::"
                      PATH_SEGMENT@87..90
                        NAME_REF@87..90
                          IDENT@87..90 "MAX"
                R_PAREN@90..91 ")"
              WHITESPACE@91..92 " "
              FAT_ARROW@92..94 "=>"
              WHITESPACE@94..95 " "
              BLOCK_EXPR@95..97
                STMT_LIST@95..97
                  L_CURLY@95..96 "{"
                  R_CURLY@96..97 "}"
              COMMA@97..98 ","
            WHITESPACE@98..103 "\n    "
            R_CURLY@103..104 "}"
        WHITESPACE@104..105 "\n"
        R_CURLY@105..106 "}"

```

</details>
